core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/ConsistencyTracker.java

```diff
@@ -0,0 +1,54 @@
+package ai.timefold.solver.core.impl.domain.variable.declarative;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+
+import ai.timefold.solver.core.impl.domain.entity.descriptor.EntityDescriptor;
+import ai.timefold.solver.core.impl.domain.solution.descriptor.SolutionDescriptor;
+
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public class ConsistencyTracker<Solution_> {
+    private final IdentityHashMap<Object, Boolean> entityToIsInconsistentMap = new IdentityHashMap<>();
+    private final HashMap<Class<?>, EntityConsistencyState<Solution_, Object>> entityClassToConsistencyStateMap =
+            new HashMap<>();
+    private boolean isFrozen;
+
+    public EntityConsistencyState<Solution_, Object>
+            getDeclarativeEntityConsistencyState(EntityDescriptor<Solution_> entityDescriptor) {
+        return entityClassToConsistencyStateMap.computeIfAbsent(entityDescriptor.getEntityClass(),
+                ignored -> new EntityConsistencyState<>(entityDescriptor, entityToIsInconsistentMap));
+    }
+
+    public void setUnknownConsistencyFromEntityShadowVariablesInconsistent(SolutionDescriptor<Solution_> solutionDescriptor,
+            Object[] entityOrFacts) {
+        var entityList = new ArrayList<>(entityOrFacts.length);
```

Comment:
Arguably doing this work via a stream can turn this into an array directly, without having the create intermediate lists.